### DISCUSSION TITLE
List pydantic types seem to be broken with v2 pydantic

### DISCUSSION THREAD

**Discussion by zoola969** [fastapi/fastapi#10331] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
####################################### v1 example

from pydantic import ConstrainedList, ConstrainedInt
from fastapi import Query


class PositiveInt(ConstrainedInt):
    gt = 0


class UniqIntList(ConstrainedList):
    min_items = 2
    item_type = PositiveInt
    unique_items = True
    __args__ = (PositiveInt,)


@app.get("/v1")
def v1_example(
    q: UniqIntList | None = Query(None),
):
    return {
        "status": "ok",
        "q": q,
    }

########################################## v2 example

from typing import Annotated
from annotated_types import Gt, MinLen
from fastapi import Query

PositiveInt = Annotated[int, Gt(0)]
UniqIntList = Annotated[set[PositiveInt], MinLen(2)]
OptionalUniqIntList = Annotated[UniqIntList | None, Query()]


@app.get("/v2")
def v2_example(
        q: UniqIntList = Query(),
        q1: UniqIntList | None = Query(None),
        q2: OptionalUniqIntList = None,
):
    return {
        "status": "ok",
        "q": q,
        "q1": q1,
        "q2": q2,
    }
```

### Description

There are several rules for arg `q`:
1. `q` should be a list
1.  Length of  `q` should be greater than `1`
1. `q` should contain only unique items
1. Items of `q` should be positive integers
1. `q` is an optional parameter, if it is not provided we set `None` as default

v2 example has 3 query parameters because I tried to make q parameter as Optional, and when I try it I get wrong validation errors and wrong swagger interface (You can see it on attached screenshots)

How v1 code works:  
1. ✔️ GET /v1 `{"status":"ok","q":null}` 
1. ✔️ GET /v1?q=1&q=2 `{"status":"ok","q":[1,2]}`
1. ✔️ GET /v1?q=1 `{"detail":[{"loc":["query","q"],"msg":"ensure this value has at least 2 items","type":"value_error.list.min_items","ctx":{"limit_value":2}}]}`
1. ✔️ GET /v1?q=1&q=0 `{"detail":[{"loc":["query","q",1],"msg":"ensure this value is greater than 0","type":"value_error.number.not_gt","ctx":{"limit_value":0}}]}`
1. ✔️ GET /v1?q=1&q=1 `{"detail":[{"loc":["query","q"],"msg":"the list has duplicated items","type":"value_error.list.unique_items"}]}`

How v2 code works:
For `q` arg:
1. ❌ GET /v2 `{"detail":{"msg":"Internal Server Error"}}` `ValueError: [TypeError("'pydantic_core._pydantic_core.PydanticUndefinedType' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]`
1. ✔️ GET /v2?q=1&q=2 `{"status":"ok","q":[1,2],"q1":null,"q2":null}`
1. ❌ GET /v2?q=1 `{"status":"ok","q":[1],"q1":null,"q2":null}` No check for list length
1. ✔️ GET /v2?q=1&q=0 `{"detail":[{"type":"greater_than","loc":["query","q",1],"msg":"Input should be greater than 0","input":"0","ctx":{"gt":0},"url":"https://errors.pydantic.dev/2.4/v/greater_than"}]}` 
1. ⚠️ GET /v2?q=1&q=1 `{"status":"ok","q":[1],"q1":null,"q2":null}` Instead of validation error we have implicit casting to set

For `q1` arg:
1. ✔️ GET /v2q=1 (set q = 1 because of previous failure)  `{"status":"ok","q":[1],"q1":null,"q2":null}` 
1. ❌ GET /v2?q=1&q1=1&q1=2 `{"detail":[{"type":"set_type","loc":["query","q1"],"msg":"Input should be a valid set","input":"2","url":"https://errors.pydantic.dev/2.4/v/set_type"}]}`
1. ❌ GET /v2?q=1&q1=1 `{"detail":[{"type":"set_type","loc":["query","q1"],"msg":"Input should be a valid set","input":"1","url":"https://errors.pydantic.dev/2.4/v/set_type"}]}`
1. ❌ GET /v2?q=1&q1=1&q1=0 `{"detail":[{"type":"set_type","loc":["query","q1"],"msg":"Input should be a valid set","input":"0","url":"https://errors.pydantic.dev/2.4/v/set_type"}]}` 
1. ❌ GET /v2?q=1&q1=1&q1=1 `{"detail":[{"type":"set_type","loc":["query","q1"],"msg":"Input should be a valid set","input":"1","url":"https://errors.pydantic.dev/2.4/v/set_type"}]}`

For `q2` arg: Same errors as for `q1` arg

So, as I see it is impossible to define custom list/set type and use it as annotation for query parameter

### Operating System

Linux

### FastAPI Version

0.103.1

### Pydantic Version

2.4.0 and 1.10.12

### Python Version

3.11.5

### Additional Context

![image](https://github.com/tiangolo/fastapi/assets/34693850/771b9cfc-f7b6-4258-9017-c2488e7ddd5b)
![image](https://github.com/tiangolo/fastapi/assets/34693850/1d494bf5-b365-48f4-8a39-589208e5ec73)

**Comment by Jan-Jasek** [comment]
Still happens on version 0.109.2, can be replicated with as simple example as

```python
@router.get("/test/")
async def test(tags: list[str] = Query(...)):
    print(tags)
```

RPC ends up with:

```
2024-02-13T08:49:03 INFO:    192.168.65.1:28855 - "GET /api/test/ HTTP/1.1" 500
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 74, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/fastapi/routing.py", line 291, in app
    raise validation_error
fastapi.exceptions.RequestValidationError: [{'type': 'list_type', 'loc': ('query', 'tags'), 'msg': 'Input should be a valid list', 'input': PydanticUndefined, 'url': 'https://errors.pydantic.dev/2.6/v/list_type'}]
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/fastapi/encoders.py", line 322, in jsonable_encoder
    data = dict(obj)
           ^^^^^^^^^
TypeError: 'pydantic_core._pydantic_core.PydanticUndefinedType' object is not iterable
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/fastapi/encoders.py", line 327, in jsonable_encoder
    data = vars(obj)
           ^^^^^^^^^
TypeError: vars() argument must have __dict__ attribute
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/usr/local/lib/python3.12/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.12/site-packages/starlette/middleware/cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 758, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 778, in app
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 299, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 79, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 75, in wrapped_app
    response = await handler(conn, exc)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/fastapi/exception_handlers.py", line 25, in request_validation_exception_handler
    content={"detail": jsonable_encoder(exc.errors())},
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/fastapi/encoders.py", line 301, in jsonable_encoder
    jsonable_encoder(
  File "/usr/local/lib/python3.12/site-packages/fastapi/encoders.py", line 287, in jsonable_encoder
    encoded_value = jsonable_encoder(
                    ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/fastapi/encoders.py", line 330, in jsonable_encoder
    raise ValueError(errors) from e
ValueError: [TypeError("'pydantic_core._pydantic_core.PydanticUndefinedType' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]
```

Basically any query list parameter that is not supplied with value (required or not) ends up throwing 500 while serving the request rather than expected 422.

@tiangolo Can we make this into an issue?

**Reply by leonidktoto** [reply]
`def test(tags: Annotated[list[str] | SkipJsonSchema[None], Query()] = None):` maybe this will help `from pydantic.json_schema import SkipJsonSchema`

**Reply by Jan-Jasek** [reply]
Well that does not help, as the contract exposes a required query params list, not an optional one that you would default to None.
So while this
`def test(tags: Annotated[list[str] | SkipJsonSchema[None], Query()] = None):
`
or
`def test(tags: list[str] | None = Query(None)):
`
technically works, it exposes a different contract.

**Reply by leonidktoto** [reply]
<img width="534" alt="Снимок экрана 2024-02-14 в 20 30 15" src="https://github.com/tiangolo/fastapi/assets/159561986/490a4cba-0dd3-4fbf-847f-951c9b24d55c">


`def test(tags: Annotated[list[int] | SkipJsonSchema[None], Query()]):` this is working

**Reply by Jan-Jasek** [reply]
Sorry, missed your response

```python
app = FastAPI()
@app.get("/test/")
def test(tags: Annotated[list[int] | SkipJsonSchema[None], Query()]):
    print(tags)
    
```
if tags are not provided, returns
`ValueError: [TypeError("'pydantic_core._pydantic_core.PydanticUndefinedType' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]`

and 500 rather than returning 422

**Comment by Harushii18** [comment]
I can confirm this error is still present

**Comment by suddenAppearance** [comment]
If anyone needs a quick patch

```python
from fastapi.encoders import ENCODERS_BY_TYPE
from pydantic_core import PydanticUndefinedType

ENCODERS_BY_TYPE[PydanticUndefinedType] = lambda o: None
```
